NPE in MimeTypes$Mutable.getMimeByExtension on startup with web-fragment with mime mapping

**Jetty version(s)**
12.1.8

**Jetty Environment**
ee10

**HTTP version**
1.1

**Java version/vendor** `(use: java -version)`

openjdk version "25.0.2" 2026-01-20
OpenJDK Runtime Environment (build 25.0.2+10-Ubuntu-125.10)
OpenJDK 64-Bit Server VM (build 25.0.2+10-Ubuntu-125.10, mixed mode, sharing)

**OS type/version**

Distributor ID:	Ubuntu
Description:	Ubuntu 25.10
Release:	25.10
Codename:	questing

**Description**

Jetty startup fails with a NPE when it tries to to scan mime types from web fragments (in our example: `eot`).

The issue seems fairly obvious when I debug into the code (see below)

* `org.eclipse.jetty.http.MimeTypes.Mutable#getMimeByExtension` gets called with `.eot`
* The implementation of `getMimeByExtension` calls `org.eclipse.jetty.util.FileID#getExtension` with `.eot`
* The implementation of `FileID.getExtension` was programmed so that it specifically returns `null` when there's a leading dot.
* Now it doesn't find a mapping and the attempt to call `equals` results in a NPE

As you can see in the screenshots below, a mapping exists for `eot`. But it doesn't get found. When I call `MimeTypes.Mutable#getMimeByExtension` with a file name  that doesn't start with a `.`, it returns the mapping as expected.

<img width="1139" height="925" alt="Image" src="https://github.com/user-attachments/assets/f620d662-0978-47ed-91a2-7963bf1d20cf" />

<img width="1076" height="857" alt="Image" src="https://github.com/user-attachments/assets/d0d00a43-7a68-4b14-8790-19dfb7a1f3a4" />

<img width="1076" height="857" alt="Image" src="https://github.com/user-attachments/assets/72c374e8-f204-47bb-bfe8-a61069c018ba" />

```
java.lang.reflect.InvocationTargetException
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:116)
    at java.lang.reflect.Method.invoke (Method.java:565)
    at org.eclipse.jetty.ee10.webapp.IterativeDescriptorProcessor.visit (IterativeDescriptorProcessor.java:81)
    at org.eclipse.jetty.ee10.webapp.IterativeDescriptorProcessor.process (IterativeDescriptorProcessor.java:69)
    at org.eclipse.jetty.ee10.webapp.WebAppContext.resolveMetaData (WebAppContext.java:1367)
    at org.eclipse.jetty.ee10.webapp.WebAppContext.startContext (WebAppContext.java:1393)
    at org.eclipse.jetty.ee10.servlet.ServletContextHandler.lambda$doStart$0 (ServletContextHandler.java:1068)
    at org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.call (ContextHandler.java:1638)
    at org.eclipse.jetty.ee10.servlet.ServletContextHandler.doStart (ServletContextHandler.java:1065)
    at org.eclipse.jetty.ee10.webapp.WebAppContext.doStart (WebAppContext.java:523)
    at org.eclipse.jetty.ee10.maven.plugin.MavenWebAppContext.doStart (MavenWebAppContext.java:312)
    at org.eclipse.jetty.util.component.AbstractLifeCycle.start (AbstractLifeCycle.java:92)
    at org.eclipse.jetty.util.component.ContainerLifeCycle.start (ContainerLifeCycle.java:170)
    at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart (ContainerLifeCycle.java:121)
    at org.eclipse.jetty.server.Handler$Abstract.doStart (Handler.java:545)
    at org.eclipse.jetty.util.component.AbstractLifeCycle.start (AbstractLifeCycle.java:92)
    at org.eclipse.jetty.util.component.ContainerLifeCycle.start (ContainerLifeCycle.java:170)
    at org.eclipse.jetty.server.Server.start (Server.java:689)
    at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart (ContainerLifeCycle.java:121)
    at org.eclipse.jetty.server.Handler$Abstract.doStart (Handler.java:545)
    at org.eclipse.jetty.server.Server.doStart (Server.java:630)
    at org.eclipse.jetty.util.component.AbstractLifeCycle.start (AbstractLifeCycle.java:92)
    at org.eclipse.jetty.maven.AbstractJettyEmbedder.doStart (AbstractJettyEmbedder.java:195)
    at org.eclipse.jetty.util.component.AbstractLifeCycle.start (AbstractLifeCycle.java:92)
    at org.eclipse.jetty.ee10.maven.plugin.JettyRunMojo.startJettyEmbedded (JettyRunMojo.java:98)
    at org.eclipse.jetty.ee10.maven.plugin.AbstractWebAppMojo.startJetty (AbstractWebAppMojo.java:483)
    at org.eclipse.jetty.ee10.maven.plugin.AbstractWebAppMojo.execute (AbstractWebAppMojo.java:461)
    at org.eclipse.jetty.ee10.maven.plugin.JettyRunMojo.execute (JettyRunMojo.java:86)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke (Method.java:565)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:41)
Caused by: java.lang.NullPointerException: Cannot invoke "String.equals(Object)" because the return value of "org.eclipse.jetty.http.MimeTypes$Mutable.getMimeByExtension(String)" is null
    at org.eclipse.jetty.ee10.webapp.StandardDescriptorProcessor.visitMimeMapping (StandardDescriptorProcessor.java:863)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke (Method.java:565)
    at org.eclipse.jetty.ee10.webapp.IterativeDescriptorProcessor.visit (IterativeDescriptorProcessor.java:81)
    at org.eclipse.jetty.ee10.webapp.IterativeDescriptorProcessor.process (IterativeDescriptorProcessor.java:69)
    at org.eclipse.jetty.ee10.webapp.WebAppContext.resolveMetaData (WebAppContext.java:1367)
    at org.eclipse.jetty.ee10.webapp.WebAppContext.startContext (WebAppContext.java:1393)
    at org.eclipse.jetty.ee10.servlet.ServletContextHandler.lambda$doStart$0 (ServletContextHandler.java:1068)
    at org.eclipse.jetty.server.handler.ContextHandler$ScopedContext.call (ContextHandler.java:1638)
    at org.eclipse.jetty.ee10.servlet.ServletContextHandler.doStart (ServletContextHandler.java:1065)
    at org.eclipse.jetty.ee10.webapp.WebAppContext.doStart (WebAppContext.java:523)
    at org.eclipse.jetty.ee10.maven.plugin.MavenWebAppContext.doStart (MavenWebAppContext.java:312)
    at org.eclipse.jetty.util.component.AbstractLifeCycle.start (AbstractLifeCycle.java:92)
    at org.eclipse.jetty.util.component.ContainerLifeCycle.start (ContainerLifeCycle.java:170)
    at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart (ContainerLifeCycle.java:121)
    at org.eclipse.jetty.server.Handler$Abstract.doStart (Handler.java:545)
    at org.eclipse.jetty.util.component.AbstractLifeCycle.start (AbstractLifeCycle.java:92)
    at org.eclipse.jetty.util.component.ContainerLifeCycle.start (ContainerLifeCycle.java:170)
    at org.eclipse.jetty.server.Server.start (Server.java:689)
    at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart (ContainerLifeCycle.java:121)
    at org.eclipse.jetty.server.Handler$Abstract.doStart (Handler.java:545)
    at org.eclipse.jetty.server.Server.doStart (Server.java:630)
    at org.eclipse.jetty.util.component.AbstractLifeCycle.start (AbstractLifeCycle.java:92)
    at org.eclipse.jetty.maven.AbstractJettyEmbedder.doStart (AbstractJettyEmbedder.java:195)
    at org.eclipse.jetty.util.component.AbstractLifeCycle.start (AbstractLifeCycle.java:92)
    at org.eclipse.jetty.ee10.maven.plugin.JettyRunMojo.startJettyEmbedded (JettyRunMojo.java:98)
    at org.eclipse.jetty.ee10.maven.plugin.AbstractWebAppMojo.startJetty (AbstractWebAppMojo.java:483)
    at org.eclipse.jetty.ee10.maven.plugin.AbstractWebAppMojo.execute (AbstractWebAppMojo.java:461)
    at org.eclipse.jetty.ee10.maven.plugin.JettyRunMojo.execute (JettyRunMojo.java:86)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:906)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:283)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:206)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke (Method.java:565)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:41)
```

**How to reproduce?**

Probably a webapp with some mime mapping in a `web-fragment.xml`, e.g. 

```
  <mime-mapping>
    <extension>eot</extension>
    <mime-type>application/vnd.ms-fontobject</mime-type>
  </mime-mapping>
```

(edit: It might only happen when starting jetty with an exploded WAR, e.g. running the jetty maven plugin on a webapp project with `resolve workspace artifacts`)

But I think the bug is fairly self-evident in the code.

--- a/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/MimeTypesTest.java
+++ b/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/MimeTypesTest.java
@@ -47,7 +47,12 @@ public class MimeTypesTest
             // Make sure that a deep path doesn't interfere
             Arguments.of("org/eclipse/jetty/Logo.png", "image/png"),
             // Make sure that path that looks like a filename doesn't interfere
-            Arguments.of("org/eclipse.jpg/jetty/Logo.png", "image/png")
+            Arguments.of("org/eclipse.jpg/jetty/Logo.png", "image/png"),
+            // Test of an undefined mime-type extensions (the extension ".bogus" shouldn't exist in mime.properties)
+            Arguments.of("foo.bogus", null),
+            // Test of dotfiles (that start with ".")
+            Arguments.of(".gitconfig", null),
+            Arguments.of(".bashrc", null)
         );
     }
 

--- a/jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/StandardDescriptorProcessor.java
+++ b/jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/StandardDescriptorProcessor.java
@@ -859,8 +859,8 @@ public class StandardDescriptorProcessor extends IterativeDescriptorProcessor
                 }
                 case WebFragment:
                 {
-                    //a web-fragment set the value, all web-fragments must have the same value
-                    if (!context.getMimeTypes().getMimeByExtension("." + extension).equals(mimeType))
+                    // a web-fragment set the value, all web-fragments must have the same value
+                    if (!mimeType.equals(context.getMimeTypes().getMimeByExtension("." + extension)))
                         throw new IllegalStateException("Conflicting mime-type " + mimeType + " for extension " + extension + " in " + descriptor.getURI());
                     break;
                 }

--- a/jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/StandardDescriptorProcessor.java
+++ b/jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/StandardDescriptorProcessor.java
@@ -860,7 +860,7 @@ public class StandardDescriptorProcessor extends IterativeDescriptorProcessor
                 case WebFragment:
                 {
                     // a web-fragment set the value, all web-fragments must have the same value
-                    if (!mimeType.equals(context.getMimeTypes().getMimeByExtension("." + extension)))
+                    if (!mimeType.equals(context.getMimeTypes().getMimeForExtension(extension)))
                         throw new IllegalStateException("Conflicting mime-type " + mimeType + " for extension " + extension + " in " + descriptor.getURI());
                     break;
                 }

--- a/jetty-ee10/jetty-ee10-webapp/src/test/java/org/eclipse/jetty/ee10/webapp/StandardDescriptorProcessorTest.java
+++ b/jetty-ee10/jetty-ee10-webapp/src/test/java/org/eclipse/jetty/ee10/webapp/StandardDescriptorProcessorTest.java
@@ -33,6 +33,7 @@ import org.eclipse.jetty.toolchain.test.FS;
 import org.eclipse.jetty.toolchain.test.MavenPaths;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDir;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDirExtension;
+import org.eclipse.jetty.util.component.LifeCycle;
 import org.eclipse.jetty.util.resource.Resource;
 import org.eclipse.jetty.util.resource.ResourceFactory;
 import org.eclipse.jetty.util.resource.Resources;
@@ -280,6 +281,10 @@ public class StandardDescriptorProcessorTest
             IllegalStateException ise = (IllegalStateException)ite.getCause();
             assertThat(ise.getMessage(), containsString("Conflicting mime-type application/vnd.ms-fontobject for extension eot in " +
                 "jar:" + fragResource.getURI().toASCIIString() + "!/META-INF/web-fragment.xml"));
+        }
+        finally
+        {
+            LifeCycle.stop(wac);
         }
     }
 

--- a/jetty-ee10/jetty-ee10-webapp/src/test/java/org/eclipse/jetty/ee10/webapp/StandardDescriptorProcessorTest.java
+++ b/jetty-ee10/jetty-ee10-webapp/src/test/java/org/eclipse/jetty/ee10/webapp/StandardDescriptorProcessorTest.java
@@ -216,7 +216,7 @@ public class StandardDescriptorProcessorTest
             """);
         wac.setDescriptor(descriptorXml.toUri().toASCIIString());
 
-        Map<String, String> CREATE_JAR_ENV = new HashMap<>();
+        final Map<String, String> CREATE_JAR_ENV = new HashMap<>();
         CREATE_JAR_ENV.put("create", "true");
 
         // Fragment one: sets the eot mime-type

--- a/jetty-ee10/jetty-ee10-webapp/src/test/java/org/eclipse/jetty/ee10/webapp/StandardDescriptorProcessorTest.java
+++ b/jetty-ee10/jetty-ee10-webapp/src/test/java/org/eclipse/jetty/ee10/webapp/StandardDescriptorProcessorTest.java
@@ -13,10 +13,14 @@
 
 package org.eclipse.jetty.ee10.webapp;
 
-import java.io.File;
 import java.lang.reflect.InvocationTargetException;
+import java.net.URI;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Arrays;
+import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
@@ -25,9 +29,13 @@ import org.eclipse.jetty.ee10.servlet.ServletHolder;
 import org.eclipse.jetty.ee10.servlet.ServletMapping;
 import org.eclipse.jetty.logging.StacklessLogging;
 import org.eclipse.jetty.server.Server;
-import org.eclipse.jetty.toolchain.test.MavenTestingUtils;
+import org.eclipse.jetty.toolchain.test.FS;
+import org.eclipse.jetty.toolchain.test.MavenPaths;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDir;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDirExtension;
+import org.eclipse.jetty.util.resource.Resource;
+import org.eclipse.jetty.util.resource.ResourceFactory;
+import org.eclipse.jetty.util.resource.Resources;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -35,13 +43,17 @@ import org.junit.jupiter.api.extension.ExtendWith;
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalToIgnoringCase;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.not;
 import static org.hamcrest.Matchers.nullValue;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 @ExtendWith(WorkDirExtension.class)
 public class StandardDescriptorProcessorTest
@@ -144,11 +156,11 @@ public class StandardDescriptorProcessorTest
     {
         //Test that the DefaultServlet mapping from webdefault-ee10.xml can be overridden in web.xml
         Path docroot = workDir.getEmptyPathDir();
-        File webXml = MavenTestingUtils.getTestResourceFile("web-redefine-mapping.xml");
+        Path webXml = MavenPaths.findTestResourceFile("web-redefine-mapping.xml");
         WebAppContext wac = new WebAppContext();
         wac.setServer(_server);
         wac.setBaseResourceAsPath(docroot);
-        wac.setDescriptor(webXml.toURI().toURL().toString());
+        wac.setDescriptor(webXml.toUri().toASCIIString());
         wac.start();
         ServletMapping[] mappings = wac.getServletHandler().getServletMappings();
         ServletMapping mapping = null;
@@ -166,15 +178,15 @@ public class StandardDescriptorProcessorTest
     }
 
     @Test
-    public void testBadDuplicateServletMappingsFromDescriptors(WorkDir workDir) throws Exception
+    public void testBadDuplicateServletMappingsFromDescriptors(WorkDir workDir)
     {
         //Test that the same mapping cannot be redefined to a different servlet in the same (non-default) descriptor
         Path docroot = workDir.getEmptyPathDir();
-        File webXml = MavenTestingUtils.getTestResourceFile("web-redefine-mapping-fail.xml");
+        Path webXml = MavenPaths.findTestResourceFile("web-redefine-mapping-fail.xml");
         WebAppContext wac = new WebAppContext();
         wac.setServer(_server);
         wac.setBaseResourceAsPath(docroot);
-        wac.setDescriptor(webXml.toURI().toURL().toString());
+        wac.setDescriptor(webXml.toUri().toASCIIString());
         wac.setThrowUnavailableOnStartupException(true);
         try (StacklessLogging ignored = new StacklessLogging(WebAppContext.class))
         {
@@ -183,14 +195,103 @@ public class StandardDescriptorProcessorTest
     }
 
     @Test
-    public void testVisitSessionConfig(WorkDir workDir) throws Exception
+    public void testDuplicateMimeTypes(WorkDir workDir) throws Exception
     {
+        // Ensure that behavior of duplicate mime-type works as intended.
         Path docroot = workDir.getEmptyPathDir();
-        File webXml = MavenTestingUtils.getTestResourceFile("web-session-config.xml");
         WebAppContext wac = new WebAppContext();
         wac.setServer(_server);
         wac.setBaseResourceAsPath(docroot);
-        wac.setDescriptor(webXml.toURI().toURL().toString());
+
+        Path descriptorXml = docroot.resolve("WEB-INF/web.xml");
+        FS.ensureDirExists(descriptorXml.getParent());
+        Files.writeString(descriptorXml, """
+            <?xml version="1.0" encoding="UTF-8"?>
+            <web-app xmlns="https://jakarta.ee/xml/ns/jakartaee"
+              xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/web-app_6_0.xsd"
+              version="6.0">
+              <name>Test WebApp</name>
+            </web-app>
+            """);
+        wac.setDescriptor(descriptorXml.toUri().toASCIIString());
+
+        Map<String, String> CREATE_JAR_ENV = new HashMap<>();
+        CREATE_JAR_ENV.put("create", "true");
+
+        // Fragment one: sets the eot mime-type
+        Path fragmentOneJar = docroot.resolve("WEB-INF/lib/frag1.jar");
+        FS.ensureDirExists(fragmentOneJar.getParent());
+        URI uriOne = URI.create("jar:" + fragmentOneJar.toUri().toASCIIString());
+        try (FileSystem zipfs = FileSystems.newFileSystem(uriOne, CREATE_JAR_ENV))
+        {
+            String fragmentXml = """
+                <?xml version="1.0" encoding="ISO-8859-1"?>
+                <web-fragment
+                   xmlns="http://java.sun.com/xml/ns/javaee"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-fragment_4_0.xsd"
+                   version="4.0">
+                  <name>Frag1Test</name>
+                  <mime-mapping>
+                    <extension>eot</extension>
+                    <mime-type>application/octet-stream</mime-type>
+                  </mime-mapping>
+                </web-fragment>
+                """;
+            Path path = zipfs.getPath("/META-INF/web-fragment.xml");
+            FS.ensureDirExists(path.getParent());
+            Files.writeString(path, fragmentXml);
+        }
+
+        // Fragment two: tries to set the eot mime-type to something difference (it should fail)
+        Path fragmentTwoJar = docroot.resolve("WEB-INF/lib/frag2.jar");
+        FS.ensureDirExists(fragmentTwoJar.getParent());
+        URI uriTwo = URI.create("jar:" + fragmentTwoJar.toUri().toASCIIString());
+        try (FileSystem zipfs = FileSystems.newFileSystem(uriTwo, CREATE_JAR_ENV))
+        {
+            String fragmentXml = """
+                <?xml version="1.0" encoding="ISO-8859-1"?>
+                <web-fragment
+                   xmlns="http://java.sun.com/xml/ns/javaee"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-fragment_4_0.xsd"
+                   version="4.0">
+                  <name>Frag2Test</name>
+                  <mime-mapping>
+                    <extension>eot</extension>
+                    <mime-type>application/vnd.ms-fontobject</mime-type>
+                  </mime-mapping>
+                </web-fragment>
+                """;
+            Path path = zipfs.getPath("/META-INF/web-fragment.xml");
+            FS.ensureDirExists(path.getParent());
+            Files.writeString(path, fragmentXml);
+        }
+
+        ResourceFactory resourceFactory = wac.getResourceFactory();
+        Resource fragResource = resourceFactory.newResource(fragmentTwoJar);
+        assertTrue(Resources.exists(fragResource));
+        wac.setThrowUnavailableOnStartupException(true);
+        try (StacklessLogging ignored = new StacklessLogging(WebAppContext.class))
+        {
+            InvocationTargetException ite = assertThrows(InvocationTargetException.class, wac::start);
+            assertThat(ite.getCause(), is(instanceOf(IllegalStateException.class)));
+            IllegalStateException ise = (IllegalStateException)ite.getCause();
+            assertThat(ise.getMessage(), containsString("Conflicting mime-type application/vnd.ms-fontobject for extension eot in " +
+                "jar:" + fragResource.getURI().toASCIIString() + "!/META-INF/web-fragment.xml"));
+        }
+    }
+
+    @Test
+    public void testVisitSessionConfig(WorkDir workDir) throws Exception
+    {
+        Path docroot = workDir.getEmptyPathDir();
+        Path webXml = MavenPaths.findTestResourceFile("web-session-config.xml");
+        WebAppContext wac = new WebAppContext();
+        wac.setServer(_server);
+        wac.setBaseResourceAsPath(docroot);
+        wac.setDescriptor(webXml.toUri().toASCIIString());
         wac.start();
         assertEquals(54, TimeUnit.SECONDS.toMinutes(wac.getSessionHandler().getMaxInactiveInterval()));
         
@@ -201,6 +302,7 @@ public class StandardDescriptorProcessorTest
         assertEquals("SPECIALSESSIONID", wac.getSessionHandler().getSessionCookie());
         
         //comment
+        //noinspection removal
         assertEquals("nocomment", wac.getSessionHandler().getSessionCookieConfig().getComment());
         assertEquals("nocomment", wac.getSessionHandler().getSessionCookieConfig().getAttribute("Comment"));
         assertEquals("nocomment", wac.getSessionHandler().getSessionComment());

--- a/jetty-ee11/jetty-ee11-webapp/src/main/java/org/eclipse/jetty/ee11/webapp/StandardDescriptorProcessor.java
+++ b/jetty-ee11/jetty-ee11-webapp/src/main/java/org/eclipse/jetty/ee11/webapp/StandardDescriptorProcessor.java
@@ -860,7 +860,7 @@ public class StandardDescriptorProcessor extends IterativeDescriptorProcessor
                 case WebFragment:
                 {
                     //a web-fragment set the value, all web-fragments must have the same value
-                    if (!mimeType.equals(context.getMimeTypes().getMimeByExtension("." + extension)))
+                    if (!mimeType.equals(context.getMimeTypes().getMimeForExtension(extension)))
                         throw new IllegalStateException("Conflicting mime-type " + mimeType + " for extension " + extension + " in " + descriptor.getURI());
                     break;
                 }

--- a/jetty-ee11/jetty-ee11-webapp/src/main/java/org/eclipse/jetty/ee11/webapp/StandardDescriptorProcessor.java
+++ b/jetty-ee11/jetty-ee11-webapp/src/main/java/org/eclipse/jetty/ee11/webapp/StandardDescriptorProcessor.java
@@ -860,7 +860,7 @@ public class StandardDescriptorProcessor extends IterativeDescriptorProcessor
                 case WebFragment:
                 {
                     //a web-fragment set the value, all web-fragments must have the same value
-                    if (!context.getMimeTypes().getMimeByExtension("." + extension).equals(mimeType))
+                    if (!mimeType.equals(context.getMimeTypes().getMimeByExtension("." + extension)))
                         throw new IllegalStateException("Conflicting mime-type " + mimeType + " for extension " + extension + " in " + descriptor.getURI());
                     break;
                 }

--- a/jetty-ee11/jetty-ee11-webapp/src/test/java/org/eclipse/jetty/ee11/webapp/StandardDescriptorProcessorTest.java
+++ b/jetty-ee11/jetty-ee11-webapp/src/test/java/org/eclipse/jetty/ee11/webapp/StandardDescriptorProcessorTest.java
@@ -33,6 +33,7 @@ import org.eclipse.jetty.toolchain.test.FS;
 import org.eclipse.jetty.toolchain.test.MavenPaths;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDir;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDirExtension;
+import org.eclipse.jetty.util.component.LifeCycle;
 import org.eclipse.jetty.util.resource.Resource;
 import org.eclipse.jetty.util.resource.ResourceFactory;
 import org.eclipse.jetty.util.resource.Resources;
@@ -280,6 +281,10 @@ public class StandardDescriptorProcessorTest
             IllegalStateException ise = (IllegalStateException)ite.getCause();
             assertThat(ise.getMessage(), containsString("Conflicting mime-type application/vnd.ms-fontobject for extension eot in " +
                 "jar:" + fragResource.getURI().toASCIIString() + "!/META-INF/web-fragment.xml"));
+        }
+        finally
+        {
+            LifeCycle.stop(wac);
         }
     }
 

--- a/jetty-ee11/jetty-ee11-webapp/src/test/java/org/eclipse/jetty/ee11/webapp/StandardDescriptorProcessorTest.java
+++ b/jetty-ee11/jetty-ee11-webapp/src/test/java/org/eclipse/jetty/ee11/webapp/StandardDescriptorProcessorTest.java
@@ -13,10 +13,14 @@
 
 package org.eclipse.jetty.ee11.webapp;
 
-import java.io.File;
 import java.lang.reflect.InvocationTargetException;
+import java.net.URI;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Arrays;
+import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
@@ -25,9 +29,13 @@ import org.eclipse.jetty.ee11.servlet.ServletHolder;
 import org.eclipse.jetty.ee11.servlet.ServletMapping;
 import org.eclipse.jetty.logging.StacklessLogging;
 import org.eclipse.jetty.server.Server;
-import org.eclipse.jetty.toolchain.test.MavenTestingUtils;
+import org.eclipse.jetty.toolchain.test.FS;
+import org.eclipse.jetty.toolchain.test.MavenPaths;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDir;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDirExtension;
+import org.eclipse.jetty.util.resource.Resource;
+import org.eclipse.jetty.util.resource.ResourceFactory;
+import org.eclipse.jetty.util.resource.Resources;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -35,13 +43,17 @@ import org.junit.jupiter.api.extension.ExtendWith;
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalToIgnoringCase;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.not;
 import static org.hamcrest.Matchers.nullValue;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 @ExtendWith(WorkDirExtension.class)
 public class StandardDescriptorProcessorTest
@@ -144,11 +156,11 @@ public class StandardDescriptorProcessorTest
     {
         //Test that the DefaultServlet mapping from webdefault-ee11.xml can be overridden in web.xml
         Path docroot = workDir.getEmptyPathDir();
-        File webXml = MavenTestingUtils.getTestResourceFile("web-redefine-mapping.xml");
+        Path webXml = MavenPaths.findTestResourceFile("web-redefine-mapping.xml");
         WebAppContext wac = new WebAppContext();
         wac.setServer(_server);
         wac.setBaseResourceAsPath(docroot);
-        wac.setDescriptor(webXml.toURI().toURL().toString());
+        wac.setDescriptor(webXml.toUri().toASCIIString());
         wac.start();
         ServletMapping[] mappings = wac.getServletHandler().getServletMappings();
         ServletMapping mapping = null;
@@ -166,15 +178,15 @@ public class StandardDescriptorProcessorTest
     }
 
     @Test
-    public void testBadDuplicateServletMappingsFromDescriptors(WorkDir workDir) throws Exception
+    public void testBadDuplicateServletMappingsFromDescriptors(WorkDir workDir)
     {
         //Test that the same mapping cannot be redefined to a different servlet in the same (non-default) descriptor
         Path docroot = workDir.getEmptyPathDir();
-        File webXml = MavenTestingUtils.getTestResourceFile("web-redefine-mapping-fail.xml");
+        Path webXml = MavenPaths.findTestResourceFile("web-redefine-mapping-fail.xml");
         WebAppContext wac = new WebAppContext();
         wac.setServer(_server);
         wac.setBaseResourceAsPath(docroot);
-        wac.setDescriptor(webXml.toURI().toURL().toString());
+        wac.setDescriptor(webXml.toUri().toASCIIString());
         wac.setThrowUnavailableOnStartupException(true);
         try (StacklessLogging ignored = new StacklessLogging(WebAppContext.class))
         {
@@ -183,14 +195,103 @@ public class StandardDescriptorProcessorTest
     }
 
     @Test
-    public void testVisitSessionConfig(WorkDir workDir) throws Exception
+    public void testDuplicateMimeTypes(WorkDir workDir) throws Exception
     {
+        // Ensure that behavior of duplicate mime-type works as intended.
         Path docroot = workDir.getEmptyPathDir();
-        File webXml = MavenTestingUtils.getTestResourceFile("web-session-config.xml");
         WebAppContext wac = new WebAppContext();
         wac.setServer(_server);
         wac.setBaseResourceAsPath(docroot);
-        wac.setDescriptor(webXml.toURI().toURL().toString());
+
+        Path descriptorXml = docroot.resolve("WEB-INF/web.xml");
+        FS.ensureDirExists(descriptorXml.getParent());
+        Files.writeString(descriptorXml, """
+            <?xml version="1.0" encoding="UTF-8"?>
+            <web-app xmlns="https://jakarta.ee/xml/ns/jakartaee"
+              xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/web-app_6_0.xsd"
+              version="6.0">
+              <name>Test WebApp</name>
+            </web-app>
+            """);
+        wac.setDescriptor(descriptorXml.toUri().toASCIIString());
+
+        Map<String, String> CREATE_JAR_ENV = new HashMap<>();
+        CREATE_JAR_ENV.put("create", "true");
+
+        // Fragment one: sets the eot mime-type
+        Path fragmentOneJar = docroot.resolve("WEB-INF/lib/frag1.jar");
+        FS.ensureDirExists(fragmentOneJar.getParent());
+        URI uriOne = URI.create("jar:" + fragmentOneJar.toUri().toASCIIString());
+        try (FileSystem zipfs = FileSystems.newFileSystem(uriOne, CREATE_JAR_ENV))
+        {
+            String fragmentXml = """
+                <?xml version="1.0" encoding="ISO-8859-1"?>
+                <web-fragment
+                   xmlns="http://java.sun.com/xml/ns/javaee"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-fragment_4_0.xsd"
+                   version="4.0">
+                  <name>Frag1Test</name>
+                  <mime-mapping>
+                    <extension>eot</extension>
+                    <mime-type>application/octet-stream</mime-type>
+                  </mime-mapping>
+                </web-fragment>
+                """;
+            Path path = zipfs.getPath("/META-INF/web-fragment.xml");
+            FS.ensureDirExists(path.getParent());
+            Files.writeString(path, fragmentXml);
+        }
+
+        // Fragment two: tries to set the eot mime-type to something difference (it should fail)
+        Path fragmentTwoJar = docroot.resolve("WEB-INF/lib/frag2.jar");
+        FS.ensureDirExists(fragmentTwoJar.getParent());
+        URI uriTwo = URI.create("jar:" + fragmentTwoJar.toUri().toASCIIString());
+        try (FileSystem zipfs = FileSystems.newFileSystem(uriTwo, CREATE_JAR_ENV))
+        {
+            String fragmentXml = """
+                <?xml version="1.0" encoding="ISO-8859-1"?>
+                <web-fragment
+                   xmlns="http://java.sun.com/xml/ns/javaee"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-fragment_4_0.xsd"
+                   version="4.0">
+                  <name>Frag2Test</name>
+                  <mime-mapping>
+                    <extension>eot</extension>
+                    <mime-type>application/vnd.ms-fontobject</mime-type>
+                  </mime-mapping>
+                </web-fragment>
+                """;
+            Path path = zipfs.getPath("/META-INF/web-fragment.xml");
+            FS.ensureDirExists(path.getParent());
+            Files.writeString(path, fragmentXml);
+        }
+
+        ResourceFactory resourceFactory = wac.getResourceFactory();
+        Resource fragResource = resourceFactory.newResource(fragmentTwoJar);
+        assertTrue(Resources.exists(fragResource));
+        wac.setThrowUnavailableOnStartupException(true);
+        try (StacklessLogging ignored = new StacklessLogging(WebAppContext.class))
+        {
+            InvocationTargetException ite = assertThrows(InvocationTargetException.class, wac::start);
+            assertThat(ite.getCause(), is(instanceOf(IllegalStateException.class)));
+            IllegalStateException ise = (IllegalStateException)ite.getCause();
+            assertThat(ise.getMessage(), containsString("Conflicting mime-type application/vnd.ms-fontobject for extension eot in " +
+                "jar:" + fragResource.getURI().toASCIIString() + "!/META-INF/web-fragment.xml"));
+        }
+    }
+
+    @Test
+    public void testVisitSessionConfig(WorkDir workDir) throws Exception
+    {
+        Path docroot = workDir.getEmptyPathDir();
+        Path webXml = MavenPaths.findTestResourceFile("web-session-config.xml");
+        WebAppContext wac = new WebAppContext();
+        wac.setServer(_server);
+        wac.setBaseResourceAsPath(docroot);
+        wac.setDescriptor(webXml.toUri().toASCIIString());
         wac.start();
         assertEquals(54, TimeUnit.SECONDS.toMinutes(wac.getSessionHandler().getMaxInactiveInterval()));
         
@@ -201,6 +302,7 @@ public class StandardDescriptorProcessorTest
         assertEquals("SPECIALSESSIONID", wac.getSessionHandler().getSessionCookie());
         
         //comment
+        //noinspection removal
         assertEquals("nocomment", wac.getSessionHandler().getSessionCookieConfig().getComment());
         assertEquals("nocomment", wac.getSessionHandler().getSessionCookieConfig().getAttribute("Comment"));
         assertEquals("nocomment", wac.getSessionHandler().getSessionComment());

--- a/jetty-ee11/jetty-ee11-webapp/src/test/java/org/eclipse/jetty/ee11/webapp/StandardDescriptorProcessorTest.java
+++ b/jetty-ee11/jetty-ee11-webapp/src/test/java/org/eclipse/jetty/ee11/webapp/StandardDescriptorProcessorTest.java
@@ -216,7 +216,7 @@ public class StandardDescriptorProcessorTest
             """);
         wac.setDescriptor(descriptorXml.toUri().toASCIIString());
 
-        Map<String, String> CREATE_JAR_ENV = new HashMap<>();
+        final Map<String, String> CREATE_JAR_ENV = new HashMap<>();
         CREATE_JAR_ENV.put("create", "true");
 
         // Fragment one: sets the eot mime-type

--- a/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/StandardDescriptorProcessor.java
+++ b/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/StandardDescriptorProcessor.java
@@ -1026,7 +1026,7 @@ public class StandardDescriptorProcessor extends IterativeDescriptorProcessor
                 case WebFragment:
                 {
                     //a web-fragment set the value, all web-fragments must have the same value
-                    if (!context.getMimeTypes().getMimeByExtension("." + extension).equals(mimeType))
+                    if (!mimeType.equals(context.getMimeTypes().getMimeByExtension("." + extension)))
                         throw new IllegalStateException("Conflicting mime-type " + mimeType + " for extension " + extension + " in " + descriptor.getURI());
                     break;
                 }

--- a/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/StandardDescriptorProcessor.java
+++ b/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/StandardDescriptorProcessor.java
@@ -1026,7 +1026,7 @@ public class StandardDescriptorProcessor extends IterativeDescriptorProcessor
                 case WebFragment:
                 {
                     //a web-fragment set the value, all web-fragments must have the same value
-                    if (!mimeType.equals(context.getMimeTypes().getMimeByExtension("." + extension)))
+                    if (!mimeType.equals(context.getMimeTypes().getMimeForExtension(extension)))
                         throw new IllegalStateException("Conflicting mime-type " + mimeType + " for extension " + extension + " in " + descriptor.getURI());
                     break;
                 }

--- a/jetty-ee9/jetty-ee9-webapp/src/test/java/org/eclipse/jetty/ee9/webapp/StandardDescriptorProcessorTest.java
+++ b/jetty-ee9/jetty-ee9-webapp/src/test/java/org/eclipse/jetty/ee9/webapp/StandardDescriptorProcessorTest.java
@@ -14,7 +14,13 @@
 package org.eclipse.jetty.ee9.webapp;
 
 import java.lang.reflect.InvocationTargetException;
+import java.net.URI;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
 import org.eclipse.jetty.ee9.servlet.DefaultServlet;
@@ -22,20 +28,28 @@ import org.eclipse.jetty.ee9.servlet.ServletHolder;
 import org.eclipse.jetty.ee9.servlet.ServletMapping;
 import org.eclipse.jetty.logging.StacklessLogging;
 import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.toolchain.test.FS;
 import org.eclipse.jetty.toolchain.test.MavenPaths;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDir;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDirExtension;
+import org.eclipse.jetty.util.resource.Resource;
+import org.eclipse.jetty.util.resource.ResourceFactory;
+import org.eclipse.jetty.util.resource.Resources;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.not;
 import static org.hamcrest.Matchers.nullValue;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 @ExtendWith(WorkDirExtension.class)
 public class StandardDescriptorProcessorTest
@@ -143,7 +157,7 @@ public class StandardDescriptorProcessorTest
         WebAppContext wac = new WebAppContext();
         wac.setServer(_server);
         wac.setBaseResourceAsPath(docroot);
-        wac.setDescriptor(webXml.toUri().toURL().toString());
+        wac.setDescriptor(webXml.toUri().toASCIIString());
         wac.start();
         ServletMapping[] mappings = wac.getServletHandler().getServletMappings();
         ServletMapping mapping = null;
@@ -169,11 +183,102 @@ public class StandardDescriptorProcessorTest
         WebAppContext wac = new WebAppContext();
         wac.setServer(_server);
         wac.setBaseResourceAsPath(docroot);
-        wac.setDescriptor(webXml.toUri().toURL().toString());
+        wac.setDescriptor(webXml.toUri().toASCIIString());
         wac.setThrowUnavailableOnStartupException(true);
         try (StacklessLogging ignored = new StacklessLogging(WebAppContext.class))
         {
             assertThrows(RuntimeException.class, () -> wac.start());
+        }
+    }
+
+    @Test
+    public void testDuplicateMimeTypes(WorkDir workDir) throws Exception
+    {
+        // Ensure that behavior of duplicate mime-type works as intended.
+        Path docroot = workDir.getEmptyPathDir();
+        WebAppContext wac = new WebAppContext();
+        wac.setServer(_server);
+        wac.setBaseResourceAsPath(docroot);
+
+        Path descriptorXml = docroot.resolve("WEB-INF/web.xml");
+        FS.ensureDirExists(descriptorXml.getParent());
+        Files.writeString(descriptorXml, """
+            <?xml version="1.0" encoding="UTF-8"?>
+            <web-app xmlns="http://xmlns.jcp.org/xml/ns/javaee"
+              xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_4_0.xsd"
+              version="4.0">
+              <name>Test WebApp</name>
+            </web-app>
+            """);
+        wac.setDescriptor(descriptorXml.toUri().toASCIIString());
+
+        Map<String, String> CREATE_JAR_ENV = new HashMap<>();
+        CREATE_JAR_ENV.put("create", "true");
+
+        // Fragment one: sets the eot mime-type
+        Path fragmentOneJar = docroot.resolve("WEB-INF/lib/frag1.jar");
+        FS.ensureDirExists(fragmentOneJar.getParent());
+        URI uriOne = URI.create("jar:" + fragmentOneJar.toUri().toASCIIString());
+        try (FileSystem zipfs = FileSystems.newFileSystem(uriOne, CREATE_JAR_ENV))
+        {
+            String fragmentXml = """
+                <?xml version="1.0" encoding="ISO-8859-1"?>
+                <web-fragment
+                   xmlns="http://java.sun.com/xml/ns/javaee"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-fragment_4_0.xsd"
+                   version="4.0">
+                  <name>Frag1Test</name>
+                  <mime-mapping>
+                    <extension>eot</extension>
+                    <mime-type>application/octet-stream</mime-type>
+                  </mime-mapping>
+                </web-fragment>
+                """;
+            Path path = zipfs.getPath("/META-INF/web-fragment.xml");
+            FS.ensureDirExists(path.getParent());
+            Files.writeString(path, fragmentXml);
+        }
+
+        // Fragment two: tries to set the eot mime-type to something difference (it should fail)
+        Path fragmentTwoJar = docroot.resolve("WEB-INF/lib/frag2.jar");
+        FS.ensureDirExists(fragmentTwoJar.getParent());
+        URI uriTwo = URI.create("jar:" + fragmentTwoJar.toUri().toASCIIString());
+        try (FileSystem zipfs = FileSystems.newFileSystem(uriTwo, CREATE_JAR_ENV))
+        {
+            String fragmentXml = """
+                <?xml version="1.0" encoding="ISO-8859-1"?>
+                <web-fragment
+                   xmlns="http://java.sun.com/xml/ns/javaee"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-fragment_4_0.xsd"
+                   version="4.0">
+                  <name>Frag2Test</name>
+                  <mime-mapping>
+                    <extension>eot</extension>
+                    <mime-type>application/vnd.ms-fontobject</mime-type>
+                  </mime-mapping>
+                </web-fragment>
+                """;
+            Path path = zipfs.getPath("/META-INF/web-fragment.xml");
+            FS.ensureDirExists(path.getParent());
+            Files.writeString(path, fragmentXml);
+        }
+
+        ResourceFactory resourceFactory = wac.getResourceFactory();
+        Resource fragResource = resourceFactory.newResource(fragmentTwoJar);
+        assertTrue(Resources.exists(fragResource));
+        wac.setThrowUnavailableOnStartupException(true);
+        try (StacklessLogging ignored = new StacklessLogging(WebAppContext.class))
+        {
+            RuntimeException re = assertThrows(RuntimeException.class, wac::start);
+            assertThat(re.getCause(), is(instanceOf(InvocationTargetException.class)));
+            InvocationTargetException ite = (InvocationTargetException)re.getCause();
+            assertThat(ite.getCause(), is(instanceOf(IllegalStateException.class)));
+            IllegalStateException ise = (IllegalStateException)ite.getCause();
+            assertThat(ise.getMessage(), containsString("Conflicting mime-type application/vnd.ms-fontobject for extension eot in " +
+                "jar:" + fragResource.getURI().toASCIIString() + "!/META-INF/web-fragment.xml"));
         }
     }
 
@@ -185,7 +290,7 @@ public class StandardDescriptorProcessorTest
         WebAppContext wac = new WebAppContext();
         wac.setServer(_server);
         wac.setBaseResourceAsPath(docroot);
-        wac.setDescriptor(webXml.toUri().toURL().toString());
+        wac.setDescriptor(webXml.toUri().toASCIIString());
         wac.start();
         assertEquals(54, TimeUnit.SECONDS.toMinutes(wac.getSessionHandler().getMaxInactiveInterval()));
         

--- a/jetty-ee9/jetty-ee9-webapp/src/test/java/org/eclipse/jetty/ee9/webapp/StandardDescriptorProcessorTest.java
+++ b/jetty-ee9/jetty-ee9-webapp/src/test/java/org/eclipse/jetty/ee9/webapp/StandardDescriptorProcessorTest.java
@@ -32,6 +32,7 @@ import org.eclipse.jetty.toolchain.test.FS;
 import org.eclipse.jetty.toolchain.test.MavenPaths;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDir;
 import org.eclipse.jetty.toolchain.test.jupiter.WorkDirExtension;
+import org.eclipse.jetty.util.component.LifeCycle;
 import org.eclipse.jetty.util.resource.Resource;
 import org.eclipse.jetty.util.resource.ResourceFactory;
 import org.eclipse.jetty.util.resource.Resources;
@@ -279,6 +280,10 @@ public class StandardDescriptorProcessorTest
             IllegalStateException ise = (IllegalStateException)ite.getCause();
             assertThat(ise.getMessage(), containsString("Conflicting mime-type application/vnd.ms-fontobject for extension eot in " +
                 "jar:" + fragResource.getURI().toASCIIString() + "!/META-INF/web-fragment.xml"));
+        }
+        finally
+        {
+            LifeCycle.stop(wac);
         }
     }
 

--- a/jetty-ee9/jetty-ee9-webapp/src/test/java/org/eclipse/jetty/ee9/webapp/StandardDescriptorProcessorTest.java
+++ b/jetty-ee9/jetty-ee9-webapp/src/test/java/org/eclipse/jetty/ee9/webapp/StandardDescriptorProcessorTest.java
@@ -213,7 +213,7 @@ public class StandardDescriptorProcessorTest
             """);
         wac.setDescriptor(descriptorXml.toUri().toASCIIString());
 
-        Map<String, String> CREATE_JAR_ENV = new HashMap<>();
+        final Map<String, String> CREATE_JAR_ENV = new HashMap<>();
         CREATE_JAR_ENV.put("create", "true");
 
         // Fragment one: sets the eot mime-type

